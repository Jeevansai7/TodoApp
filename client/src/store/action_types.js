export const FETCH_DATA_PENDING = 'FETCH_DATA_PENDING';
export const FETCH_DATA_SUCCESS = 'FETCH_DATA_SUCCESS';
export const FETCH_DATA_ERROR = 'FETCH_DATA_ERROR';

export const CHANGES_IN_RECORDS_PENDING ='CHANGES_IN_RECORDS_PENDING';
export const CHANGES_IN_RECORDS_SUCCESS ='CHANGES_IN_RECORDS_SUCCESS';
export const CHANGES_IN_RECORDS_ERROR ='CHANGES_IN_RECORDS_ERROR';
