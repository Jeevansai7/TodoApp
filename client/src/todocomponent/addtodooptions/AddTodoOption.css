.addtodoapp{
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    width: 55%;
    justify-content: center;
    position: relative;
    left: 20%;
    border-radius: 50px;
    border-color: black black lime  lime;
    border-width: 4px;
    border-style: solid;
}

.addtodoapp input{
    width: 90%;
    height: 6vh;
    border-radius: 50px;
    border: none;
    outline: none;
    caret-color: black;
    font-size: large;
    font-weight: bolder;
    padding-left: 10px;
}

.addtodoapp button{
    color: black;
   background: white;
    width: 10%;
    font-size: larger;
    font-weight: bolder;
    border-radius: 50px;
    border: none;
    height: 6vh;
}

.addtodoapp button:hover{
     background-color: black; 
    color: lime;
    border-radius: 50px;
    cursor: pointer;
}

.addtodoapp:hover{
    animation: borderstyle 1s ease 0s infinite normal ;
}

@keyframes borderstyle {
    0% {border-color: black black lime  lime}
    25%{border-color: lime black black lime}
    50% {border-color: lime lime black black}
    100% {border-color: black black lime  lime}
}