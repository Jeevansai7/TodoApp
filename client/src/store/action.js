import {FETCH_DATA_PENDING,FETCH_DATA_SUCCESS,FETCH_DATA_ERROR,CHANGES_IN_RECORDS_PENDING, CHANGES_IN_RECORDS_SUCCESS, CHANGES_IN_RECORDS_ERROR} from './action_types'
import { getData } from '../services/getDataService';
import axios from 'axios';
import { deleteRecord } from '../services/deleteRecordService';
import { editRecord } from '../services/editRecordService';
import { addRecordService } from '../services/addRecordService';

export const fetchData=()=>{
    return async dispatch =>{
        try{
            dispatch (serviceActionPending(FETCH_DATA_PENDING));
            let response = await getData(instance);
            if(response){
                dispatch(serviceActionSuccess(FETCH_DATA_SUCCESS,response));
            }else{
                dispatch(serviceActionError(FETCH_DATA_ERROR,"Unknown error"));
            }
        } 
        catch(e){
            console.error(e);
            dispatch(serviceActionError(FETCH_DATA_ERROR,"Unknown error"));
        }
    }

}

export const deleteRecordById = (id,fetchTodoData) => {
    return async dispatch => {
        try {
            dispatch(serviceActionPending(CHANGES_IN_RECORDS_PENDING));
            const response = await deleteRecord(instance, id);
            await fetchTodoData();
            if (response) {
                dispatch(serviceActionSuccess(CHANGES_IN_RECORDS_SUCCESS));
            } else {
                dispatch(serviceActionError(CHANGES_IN_RECORDS_ERROR, "Unknown error"));
            }
        }
        catch (e) {
            console.error(e);
            dispatch(serviceActionError(CHANGES_IN_RECORDS_ERROR, "Unknown error"));
        }
    }

}

export const editRecordById = (id,task,fetchTodoData) => {
    return async dispatch => {
        try {
            dispatch(serviceActionPending(CHANGES_IN_RECORDS_PENDING));
            const response = await editRecord(instance, id,task);
            await fetchTodoData();
            if (response) {
                dispatch(serviceActionSuccess(CHANGES_IN_RECORDS_SUCCESS));
            } else {
                dispatch(serviceActionError(CHANGES_IN_RECORDS_ERROR, "Unknown error"));
            }
        }
        catch (e) {
            console.error(e);
            dispatch(serviceActionError(CHANGES_IN_RECORDS_ERROR, "Unknown error"));
        }
    }

}

export const addRecord = (task,fetchTodoData) => {
    return async dispatch => {
        try {
            dispatch(serviceActionPending(CHANGES_IN_RECORDS_PENDING));
            const response = await addRecordService(instance,task);
            await fetchTodoData();
            if (response) {
                dispatch(serviceActionSuccess(CHANGES_IN_RECORDS_SUCCESS));
            } else {
                dispatch(serviceActionError(CHANGES_IN_RECORDS_ERROR, "Unknown error"));
            }
        }
        catch (e) {
            console.error(e);
            dispatch(serviceActionError(CHANGES_IN_RECORDS_ERROR, "Unknown error"));
        }
    }

}

function serviceActionPending(type) {
    return {
        type: type
    }
}

function serviceActionSuccess(type, data) {
    return {
        type: type,
        data: data
    }
}

function serviceActionError(type, error) {
    return {
        type: type,
        error: error
    }
}

const instance = axios.create({
    "baseURL":'https://todoappserver-ei5w.onrender.com/',
    "headers":{
        "Content-Type":'application/json'
    }
})

