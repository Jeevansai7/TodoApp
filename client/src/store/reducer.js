import { CHANGES_IN_RECORDS_ERROR, CHANGES_IN_RECORDS_PENDING, CHANGES_IN_RECORDS_SUCCESS,FETCH_DATA_ERROR, FETCH_DATA_PENDING, FETCH_DATA_SUCCESS } from "./action_types";

const initialState = {
   pending:false,
   todoData:[],
}


export default function TodoReducer (state=initialState,action){
    switch(action.type){
        case FETCH_DATA_PENDING:
            return Object.assign({},state,{
                pending:true,
            });
        case FETCH_DATA_SUCCESS:
            return Object.assign({},state,{
                pending:false,
                todoData:action.data
            });
        case FETCH_DATA_ERROR:
            return Object.assign({},state,{
                pending:false,
                todoData:[]
            });
        case CHANGES_IN_RECORDS_PENDING:
            return Object.assign({},state,{
                pending:true,
            });
        case CHANGES_IN_RECORDS_SUCCESS:
            return Object.assign({},state,{
                pending:false,
            });
        case CHANGES_IN_RECORDS_ERROR:
            return Object.assign({},state,{
                pending:false,
            });
        default:
            return state;
    }
}