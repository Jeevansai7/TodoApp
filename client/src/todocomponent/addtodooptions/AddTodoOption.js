import React,{useRef} from 'react'
import './AddTodoOption.css'

const AddTodoOption =(props)=>{

    const {clickHandler}=props
    const inputRef=useRef("")

    const onAdd=()=>{
        if(inputRef.current.value!=''){
        clickHandler(inputRef.current.value); 
        inputRef.current.value=''
        }
    }

    return(
        <div className='addtodoapp'>
        <input type="search" ref={inputRef} onKeyDown={event=>{if(event.key=='Enter'){onAdd()}}}/>
        <button onClick={()=>onAdd()}>Add</button>
        </div>
    )
};
export default AddTodoOption;