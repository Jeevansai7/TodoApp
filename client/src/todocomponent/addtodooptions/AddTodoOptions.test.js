import React from 'react';
import {render,screen} from '../../test/test-utils';
import userEvent from '@testing-library/user-event'
import TestRenderer from 'react-test-renderer';
import AddTodoOption from './AddTodoOption';
import store from '../../store/store'
import { addRecord } from '../../store/action';
import * as allActions from '../../store/action';

describe('AddTodoOption Component',()=>{
    
    test('input box renders',()=>{
        render(<AddTodoOption />);

        const inputElement=screen.getByRole('searchbox');

        expect(inputElement).toBeInTheDocument();
    });

    test('user should be able to see add button',()=>{
        
        render(<AddTodoOption />)

        const addButton=screen.getByRole('button',{name:'Add'});
       
        expect(addButton).toBeInTheDocument();

    });

    test('user should be able to type the value inside input box ',()=>{
        
        render(<AddTodoOption />)

        const inputElement=screen.getByRole('searchbox');
       
        expect(inputElement).toHaveValue('');

        userEvent.type(inputElement, 'Risk');
        expect(inputElement).toHaveValue('Risk');

        userEvent.type(inputElement,'100');
        expect(inputElement).toHaveValue('Risk100');

    });

    test('click handler should not be called when input box is empty',()=>{
        const addHandler=jest.fn()
        render(<AddTodoOption clickHandler={addHandler}/>)

        const inputElement=screen.getByRole('searchbox');
       
        expect(inputElement).toHaveValue('');

        userEvent.type(inputElement,'{enter}');
        expect(addHandler).not.toHaveBeenCalled();

    });

    test('click handler should be called when input box is not empty',()=>{
        const fetchTodoData =async() =>{
            await dispatch(fetchData())
        }
        const useDispatchMock = jest.spyOn(allActions,'addRecord')
        
        const addHandler=jest.fn((userValue)=>store.dispatch(addRecord(userValue,fetchTodoData)))
        render(<AddTodoOption clickHandler={addHandler}/>)

        const inputElement=screen.getByRole('searchbox');
       
        expect(inputElement).toHaveValue('');

        userEvent.type(inputElement, 'need to check exercise');
        expect(inputElement).toHaveValue('need to check exercise');

        userEvent.type(inputElement,'{enter}');

        expect(addHandler).toHaveBeenCalled();
        expect(addHandler).toHaveBeenCalledWith('need to check exercise');
        expect(useDispatchMock).toHaveBeenCalled()
        expect(useDispatchMock).toHaveBeenCalledWith('need to check exercise',fetchTodoData)
       
    });

    test('renders correctly', () => {
        const addHandler=jest.fn((userValue)=>store.dispatch(addRecord(userValue,fetchTodoData)))
        const tree = TestRenderer.create(<AddTodoOption clickHandler={addHandler} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
});