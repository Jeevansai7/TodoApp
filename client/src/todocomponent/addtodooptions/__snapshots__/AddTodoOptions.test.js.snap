// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodoOption Component renders correctly 1`] = `
<div
  className="addtodoapp"
>
  <input
    onKeyDown={[Function]}
    type="search"
  />
  <button
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;
